TextFieldEffects/TextFieldEffects/HoshiTextField.swift

```diff
@@ -12,178 +12,226 @@ import UIKit
+                      self.placeholderLabel.frame.origin = self.activePlaceholderPoint
+                     }), completion: { _ in
+                      self.animationCompletionHandler?(.textEntry)
+
+      })
     }
-    
-    // MARK: - Overrides
-    
-    override open func editingRect(forBounds bounds: CGRect) -> CGRect {
-        return bounds.offsetBy(dx: textFieldInsets.x, dy: textFieldInsets.y)
+  }
+
+  override open func animateViewsForTextDisplay() {
+    if text!.isEmpty {
+      UIView.animate(withDuration: 0.35, delay: 0.0,
+                     usingSpringWithDamping: 0.8,
+                     initialSpringVelocity: 2.0,
+                     options: UIViewAnimationOptions.beginFromCurrentState,
+                     animations: ({
+                      self.placeholderLabel.transform = CGAffineTransform(scaleX: 1.0, y: 1.0)
+                      guard let fontSize = self.font?.pointSize else { return }
+                      guard let fontName = self.font?.fontName else { return }
+                      self.placeholderLabel.font = UIFont(name: fontName,
+                                                          size: fontSize)
+                      self.placeholderLabel.sizeToFit()
+                      self.placeholderLabel.frame.origin = self.inactivePlaceholderPoint
+                     }), completion: { [weak self] _  in
+                      guard let strongSelf = self else { return }
+                      strongSelf.animationCompletionHandler?(.textDisplay)
+      })
+    }
+  }
+
+  // MARK: - Private
+  private func initPlaceholderFont() {
+    guard let fontSize = self.font?.pointSize else { return }
+    guard let fontName = self.font?.fontName else { return }
+    self.placeholderLabel.font = UIFont(name: fontName,
+                                        size: fontSize)
+
+  }
+
+  private func configurePlaceholderLabelFrame() {
+    placeholderLabel.frame = frame.insetBy(dx: placeholderInsets.x, dy: placeholderInsets.y)
+    placeholderLabel.frame.origin = self.inactivePlaceholderPoint
+  }
+
+  private func configurePlaceholderFont() {
+    if let font = UIFont(name: "Roboto-Regular", size: 12) {
+      placeholderFont = font
     }
-    
-    override open func textRect(forBounds bounds: CGRect) -> CGRect {
-        return bounds.offsetBy(dx: textFieldInsets.x, dy: textFieldInsets.y)
+  }
+
+  private func updateBorder() {
+    inactiveBorderLayer.frame = rectForBorder(borderThickness.inactive)
+    inactiveBorderLayer.backgroundColor = borderInactiveColor?.cgColor
+
+    activeBorderLayer.frame = rectForBorder(borderThickness.active)
+    activeBorderLayer.backgroundColor = borderActiveColor?.cgColor
+    activeBorderLayer.isHidden = true
+  }
+
+  private func updatePlaceholder() {
+    placeholderLabel.text = placeholder
+    placeholderLabel.textColor = placeholderColor
+    placeholderLabel.sizeToFit()
+
+    if isFirstResponder || text!.isNotEmpty {
```

Comment:
Are you sure to force unwrap the text over here? Can try `guard` or `if let` it ?